chainermn/communicators/communicator_base.py

```diff
@@ -0,0 +1,263 @@
+
+    '''
+
+    def __init__(self):
+        pass
+
+    @property
+    def rank(self):
+        '''Rank (process id in the cluster) of this process in integer.'''
+        raise NotImplementedError()
+
+    @property
+    def intra_rank(self):
+        '''Intra rank (process id in the machine) of this process integer'''
+        raise NotImplementedError()
+
+    @property
+    def size(self):
+        '''Number of nodes of the cluster'''
+        raise NotImplementedError()
+
+    @abstractmethod
+    def alltoall(self, xs):
+        '''All-to-all implementation for ndarray
+
+        Args:
+            xs (tuple of numpy.ndarray)
+
+        Returns:
+            ys (tuple of numpy.ndarray):
+                Received arrays. The length of tuple equals to
+                the communicator size.
+
+        '''
+        raise NotImplementedError()
+
+    # on ndarrays and such
+    @abstractmethod
+    def send(self, data, dest, tag):
+        '''Sends an ndarray to destination
+
+        Receiver must invoke ``recv()`` to wait for the message.
+
+        Args:
+            data: data to be sent (tuple, list or raw numpy/cupy array)
+            dest (int): Rank of the destination process
+            tag (int): The tag to identify the message
+
+        '''
+        raise NotImplementedError()
+
+    @abstractmethod
+    def recv(self, source, tag):
+        '''Receives an ndarray from source.
+
+        To receive the message, sender must send the data.
+
+        Args:
+            source (int): Rank of the source process
+            tag (int): The tag to specifically receive the message
+
+        Returns:
+            The data sent from source process
+
+        '''
+        raise NotImplementedError()
+
+    @abstractmethod
+    def bcast(self, data, max_buf_len=None, root=0):
+        '''Broadcasts an ndarray from root process to all processes
+
+        Args:
+            data (numpy.array): for root process, the data to broadcast.
+                For non-root processes, this argument is ignored.
+            max_buf_len (int): Length of send buffer.
+            root (int): the process who has the data to broadcast.
+
+        Returns:
+            ys (numpy.ndarray) : The data sent from root process
+
+        '''
+        raise NotImplementedError()
+
+    @abstractmethod
+    def gather(self, data, root=0):
+        '''Gathers an ndarray from all processes to root process
+
+        Args:
+            data (ndarray, or scaler): for root process this is ignored. For
+                For non-root processes, the data to send to root process.
+            root (int): rank of the process who receives the data.
+
+        Returns:
+            For root process, the ndarray sent from non-root processes.
+            For non-root processes, what?
+
+        '''
+        raise NotImplementedError()
+
+    # TODO(kuenishi): implement this function from MpiCommunicatorBase #234
+    # @abstractmethod
+    def allreduce(self, data, op=None):
```

Comment:
I don't think we should introduce `op` now, as none of implementations support it. When necessary, adding arguments with default values afterwards is totally OK.